# 05 — Money: LTV, Profit Curve, Optimizer & Sensitivity
**RetainIQ India (Days 9-11).** The decision chain: probabilities → rupees → a budget-constrained contact list → stress tests.

> All monetary figures are **(simulation-based estimates)**. The dataset has ARPU and tenure but *no* cost, margin, offer or acceptance data — those are **declared assumptions** in `src/economics.py`.

In [1]:
import sys, pandas as pd, json
sys.path.insert(0,'../src')
from economics import ECON
print(json.dumps(ECON.as_dict(), indent=2))

{
  "gross_margin": 0.6,
  "monthly_discount_rate": 0.008,
  "horizon_months": 24,
  "offer_cost_inr": 120.0,
  "acceptance_rate": 0.35,
  "budget_inr": 100000.0
}


## 1. LTV — survival-weighted, not naive
`LTV = Σ ARPU × margin × S_contract(t) / (1+r)^t`. Using the Day-5 Kaplan-Meier curves means a month-to-month customer's future months are discounted by their *real* probability of still being there.

In [2]:
import ltv
vals = ltv.build()
vals.groupby('contract_type').agg(n=('customer_id','size'), mean_ltv=('ltv_inr','mean'),
    mean_churn=('churn_proba','mean'), mean_value_at_risk=('value_at_risk_inr','mean')).round(2)

05:59:07 | INFO    | model | leakage-free split: train=5634 test=1409 (stratified, seed=42)


05:59:07 | INFO    | model | Logistic (calibrated): {'roc_auc': 0.8441, 'pr_auc': 0.6595, 'brier': 0.1367, 'accuracy': 0.7999, 'precision': 0.6554, 'recall': 0.5187, 'f1': 0.5791}


05:59:07 | INFO    | model | Brier raw=0.1367 -> calibrated=0.1367


05:59:08 | INFO    | model | RandomForest (challenger): {'roc_auc': 0.846, 'pr_auc': 0.6705, 'brier': 0.1357, 'accuracy': 0.7999, 'precision': 0.6523, 'recall': 0.5267, 'f1': 0.5828}


05:59:08 | INFO    | model | saved primary model -> models/churn_model.pkl


05:59:08 | INFO    | model | survival-discount factor [Month-to-month] = 15.56 effective months


05:59:08 | INFO    | model | survival-discount factor [One year      ] = 21.56 effective months


05:59:08 | INFO    | model | survival-discount factor [Two year      ] = 21.76 effective months


05:59:09 | INFO    | model | LTV built for 7,043 customers -> data/processed/customer_value.parquet


05:59:09 | INFO    | model | mean LTV ₹708 | median ₹714 | total portfolio ₹0.5 Cr


05:59:09 | INFO    | model | total value at risk = ₹13.6 L  (simulation-based)


05:59:09 | INFO    | model | customers with positive expected_net at ₹120 offer: 1730 / 7043


05:59:09 | INFO    | model | LTV by contract:
                   n  mean_ltv  mean_churn  mean_var
contract_type                                       
Month-to-month  3875    619.88        0.43    291.46
One year        1473    841.56        0.11    116.92
Two year        1695    793.34        0.03     33.12


,n,mean_ltv,mean_churn,mean_value_at_risk
contract_type,,,,
Month-to-month,3875,619.88,0.43,291.46
One year,1473,841.56,0.11,116.92
Two year,1695,793.34,0.03,33.12


## 2. The rupee profit curve
For each contact threshold: benefit − cost. This exposes the **cost-sensitive threshold**.

In [3]:
import profit_curve as PC
pc = PC.run()
print('contact-everyone net : Rs', pc['contact_everyone']['net_retained_inr'])
print('profit-max threshold :', pc['optimal_threshold'])
print('  -> contact', pc['optimal']['n_contacted'], '| net Rs', pc['optimal']['net_retained_inr'])
print('per-customer +EV ceiling: Rs', pc['theoretical_ceiling_inr'])

05:59:09 | INFO    | model | Contact-everyone baseline: n=7043 net=₹-369938


05:59:09 | INFO    | model | PROFIT-MAXIMISING threshold = 0.50 | contact 1529 | net ₹74527


05:59:09 | INFO    | model | Theoretical ceiling (contact every +EV customer): ₹89549


05:59:09 | INFO    | model | Uplift vs contact-everyone: +120.1%  (simulation-based)


05:59:09 | INFO    | model | At the naive 0.50 cut: contact 1529 | net ₹74527 (leaves ₹0 on the table)


05:59:09 | INFO    | model | saved reports/profit_curve.json + reports/figures/profit_curve.png


contact-everyone net : Rs -369938.06
profit-max threshold : 0.5
  -> contact 1529 | net Rs 74527.17
per-customer +EV ceiling: Rs 89549.15


![profit curve](../reports/figures/profit_curve.png)

**Key insight.** The profit-max *global* threshold lands near 0.50 only by coincidence of these assumptions. Break-even churn probability actually varies **0.22 → 1.96** across customers because it depends on LTV. So **no single threshold is optimal** — which is exactly why we need an optimizer, not a cut-off.

## 3. The budget-constrained optimizer  ★ the thesis
0/1 knapsack: maximise Σ(benefit − cost) subject to Σcost ≤ budget. Solved greedily by ROI-per-rupee, verified against a DP bound.

In [4]:
import optimizer as O
res = O.run()
pd.DataFrame(res['results']).T[['n_contacted','spend_inr','net_retained_inr']]

05:59:09 | INFO    | model | Offer assignment (7,043 customers):


05:59:09 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:09 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:09 | INFO    | model | optimizer_roi        n= 1349 spend=₹   100000 net=₹     63133


05:59:09 | INFO    | model | rank_by_probability  n=  843 spend=₹    99980 net=₹     50934


05:59:09 | INFO    | model | random               n= 1993 spend=₹    99980 net=₹    -18351


05:59:09 | INFO    | model | contact_everyone     n= 7043 spend=₹   359060 net=₹    -60009 (ignores budget)


05:59:09 | INFO    | model | budget utilisation: 100.0% of ₹100000


05:59:09 | INFO    | model | uplift vs rank_by_probability     +23.9%  (simulation-based)


05:59:09 | INFO    | model | uplift vs random                 +444.0%  (simulation-based)


05:59:09 | INFO    | model | uplift vs contact_everyone       +205.2%  (simulation-based)


05:59:09 | INFO    | model | DP knapsack bound ₹63133 vs greedy ₹63133 -> gap 0.000%


05:59:09 | INFO    | model | selected offer mix:
                     n    spend      net
offer                                   
Bill discount       50  11000.0   4835.0
Courtesy call      836  33440.0  23688.0
Protection bundle  463  55560.0  34610.0


05:59:09 | INFO    | model | saved reports/optimizer_result.json + reports/contact_list.csv


,n_contacted,spend_inr,net_retained_inr
optimizer_roi,1349.0,100000.0,63132.68
rank_by_probability,843.0,99980.0,50934.34
random,1993.0,99980.0,-18350.88
contact_everyone,7043.0,359060.0,-60009.06


In [5]:
print('budget utilisation :', res['budget_utilisation_pct'], '%')
print('greedy vs DP gap   :', res['greedy_gap_pct'], '%')
print('uplift vs baselines:', json.dumps(res['uplift_pct'], indent=2))

budget utilisation : 100.0 %
greedy vs DP gap   : 0.0 %
uplift vs baselines: {
  "rank_by_probability": 23.9,
  "random": 444.0,
  "contact_everyone": 205.2
}


**Headline `(simulation-based)`:** under a binding ₹1,00,000 budget the optimizer retains **₹63,133 net — +23.9% vs ranking by churn probability alone**, and +205% vs contacting everyone (which *loses* ₹60,009). Greedy matches the DP optimum exactly.

## 4. Sensitivity — which assumption breaks the recommendation first?

In [6]:
import sensitivity as SN
sn = SN.run()
print('break-even acceptance scale     :', sn['break_even_acceptance_scale'])
print('=> protection-bundle acceptance  :', sn['break_even_protection_acceptance_pct'], '% (baseline assumes 35%)')
pd.DataFrame(sn['one_way']['offer_cost_multiplier'])

05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:10 | INFO    | model | baseline net retained = ₹63133


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call         92       0.77   916.28     2.32      3680.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 92 / 7043


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Courtesy call           1058       0.65   842.38     8.53     42320.0
Protection bundle         44       0.80   882.82    28.12      5280.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 1102 / 7043


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount              3       0.78   969.98    53.57       660.0
Courtesy call           1430       0.56   852.87    15.64     57200.0
Protection bundle        340       0.73   799.83    43.27     40800.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 1773 / 7043


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            378       0.64   919.48    97.04     83160.0
Courtesy call           1295       0.41   889.73    21.89     51800.0
Protection bundle        747       0.67   752.51    90.41     89640.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 2420 / 7043


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            646       0.61   909.04   121.97    142120.0
Courtesy call           1285       0.36   880.23    21.11     51400.0
Protection bundle        777       0.65   743.12   117.22     93240.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 2708 / 7043


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            944       0.58   903.41   153.04    207680.0
Courtesy call           1285       0.33   853.64    19.95     51400.0
Protection bundle        737       0.63   727.86   135.78     88440.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 2966 / 7043


05:59:10 | INFO    | model | acceptance sweep: {0.4: 214.0, 0.6: 10259.0, 0.8: 37234.0, 1.0: 63133.0, 1.2: 87481.0, 1.4: 111292.0, 1.6: 130865.0}


05:59:10 | INFO    | model | Offer assignment (7,043 customers):


05:59:10 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount           1521       0.56   889.60   107.95    167310.0
Courtesy call           1260       0.29   814.10     9.52     25200.0
Protection bundle        594       0.57   673.78    71.40     35640.0


05:59:10 | INFO    | model | eligible (+EV under best offer): 3375 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            554       0.61   914.33    82.94     91410.0
Courtesy call           1299       0.38   876.96    15.83     38970.0
Protection bundle        774       0.66   749.61    83.10     69660.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 2627 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Courtesy call           1244       0.62   836.12    16.15     74640.0
Protection bundle        129       0.77   853.11    49.19     23220.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 1373 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call        534        0.7   873.12    11.31     42720.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 534 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:11 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:11 | INFO    | model | offer-cost sweep: {0.5: 143871.0, 0.75: 94486.0, 1.0: 63133.0, 1.5: 26433.0, 2.0: 6039.0, 3.0: 0.0}


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Courtesy call           1244       0.62   557.42    10.77     49760.0
Protection bundle        129       0.77   568.74    32.79     15480.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 1373 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount             15       0.77   798.41    56.86      3300.0
Courtesy call           1424       0.54   718.46    16.57     56960.0
Protection bundle        411       0.72   656.10    45.48     49320.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 1850 / 7043


05:59:11 | INFO    | model | Offer assignment (7,043 customers):


05:59:11 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:11 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            346       0.65  1075.53    92.12     76120.0
Courtesy call           1288       0.42  1042.01    21.61     51520.0
Protection bundle        740       0.67   877.23    85.20     88800.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2374 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            554       0.61  1219.11   110.59    121880.0
Courtesy call           1299       0.38  1169.28    21.11     51960.0
Protection bundle        774       0.66   999.48   110.80     92880.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2627 / 7043


05:59:12 | INFO    | model | margin sweep: {0.4: 17622.0, 0.5: 42476.0, 0.6: 63133.0, 0.7: 83172.0, 0.8: 103963.0}


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:12 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:12 | INFO    | model | budget sweep: {25000: 24075.0, 50000: 40602.0, 100000: 63133.0, 150000: 77105.0, 200000: 78490.0, 400000: 78490.0}


05:59:12 | INFO    | model | Offer assignment (7,043 customers):


05:59:12 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:12 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call        565        0.7   870.66     5.79     22600.0


05:59:13 | INFO    | model | eligible (+EV under best offer): 565 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:13 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call         49       0.79   921.15     1.57      1960.0


05:59:13 | INFO    | model | eligible (+EV under best offer): 49 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:13 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.05        40.0


05:59:13 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:13 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:13 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:13 | INFO    | model | Offer assignment (7,043 customers):


05:59:13 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:13 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:14 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:14 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:14 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.02        40.0


05:59:14 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:14 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:14 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:14 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:14 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:14 | INFO    | model | Offer assignment (7,043 customers):


05:59:14 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:14 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:15 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:15 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:15 | INFO    | model | Offer assignment (7,043 customers):


05:59:15 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:15 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:16 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:16 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:16 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:16 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:16 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:16 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:16 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:16 | INFO    | model | Offer assignment (7,043 customers):


05:59:16 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:16 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:17 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:17 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:17 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:17 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:17 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.01        40.0


05:59:17 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:17 | INFO    | model | BREAK-EVEN acceptance scale = 0.350 (i.e. protection-bundle acceptance ≈ 12.2%)


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount              3       0.78   969.98    26.78       330.0
Courtesy call           1430       0.56   852.87     7.82     28600.0
Protection bundle        340       0.73   799.83    21.63     20400.0


05:59:17 | INFO    | model | eligible (+EV under best offer): 1773 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call         92       0.77   916.28     2.32      3680.0


05:59:17 | INFO    | model | eligible (+EV under best offer): 92 / 7043


05:59:17 | INFO    | model | Offer assignment (7,043 customers):


05:59:17 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:17 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:18 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:18 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            646       0.61   909.04    60.99     71060.0
Courtesy call           1285       0.36   880.23    10.55     25700.0
Protection bundle        777       0.65   743.12    58.61     46620.0


05:59:18 | INFO    | model | eligible (+EV under best offer): 2708 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Courtesy call           1345       0.60   835.89    12.02     53800.0
Protection bundle        153       0.76   844.25    37.73     18360.0


05:59:18 | INFO    | model | eligible (+EV under best offer): 1498 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call        348       0.72   892.61     7.15     20880.0


05:59:18 | INFO    | model | eligible (+EV under best offer): 348 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call          1       0.81   941.04     0.03        80.0


05:59:18 | INFO    | model | eligible (+EV under best offer): 1 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:18 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:18 | INFO    | model | Offer assignment (7,043 customers):


05:59:18 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount           1521       0.56   889.60   107.95    167310.0
Courtesy call           1260       0.29   814.10     9.52     25200.0
Protection bundle        594       0.57   673.78    71.40     35640.0


05:59:18 | INFO    | model | eligible (+EV under best offer): 3375 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Courtesy call           1244       0.62   836.12    16.15     74640.0
Protection bundle        129       0.77   853.11    49.19     23220.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 1373 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call        534        0.7   873.12    11.31     42720.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 534 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:19 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount           2163       0.54   870.67   151.33    237930.0
Courtesy call           1206       0.25   806.16    10.44     24120.0
Protection bundle        392       0.46   578.51    54.59     23520.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 3761 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            513       0.62   918.73   106.77    112860.0
Courtesy call           1288       0.39   877.60    21.37     51520.0
Protection bundle        771       0.66   749.63   105.37     92520.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 2572 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount             29       0.76   951.65    93.24      9570.0
Courtesy call           1399       0.52   870.63    26.39     83940.0
Protection bundle        484       0.71   775.12    71.34     87120.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 1912 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Courtesy call           1210       0.63   836.26    20.21     96800.0
Protection bundle        109       0.78   855.86    62.01     26160.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 1319 / 7043


05:59:19 | INFO    | model | Offer assignment (7,043 customers):


05:59:19 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call        223       0.74   895.91     9.56     26760.0


05:59:19 | INFO    | model | eligible (+EV under best offer): 223 / 7043


05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount           2545       0.51   868.57   190.08    279950.0
Courtesy call           1125       0.20   835.90    10.80     22500.0
Protection bundle        376       0.44   495.64    51.16     22560.0


05:59:20 | INFO    | model | eligible (+EV under best offer): 4046 / 7043


05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            944       0.58   903.41   153.04    207680.0
Courtesy call           1285       0.33   853.64    19.95     51400.0
Protection bundle        737       0.63   727.86   135.78     88440.0


05:59:20 | INFO    | model | eligible (+EV under best offer): 2966 / 7043


05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            217       0.68   928.57   123.06     71610.0
Courtesy call           1322       0.45   894.57    31.93     79320.0
Protection bundle        687       0.68   753.74   107.63    123660.0


05:59:20 | INFO    | model | eligible (+EV under best offer): 2226 / 7043


05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount              3       0.78   969.98   107.13      1320.0
Courtesy call           1430       0.56   852.87    31.28    114400.0
Protection bundle        340       0.73   799.83    86.53     81600.0


05:59:20 | INFO    | model | eligible (+EV under best offer): 1773 / 7043


05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
               customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                            
Courtesy call        717       0.68   862.65    20.48     86040.0


05:59:20 | INFO    | model | eligible (+EV under best offer): 717 / 7043


05:59:20 | INFO    | model | two-way grid (net retained ₹):
          cost x0.5  cost x1.0  cost x1.5  cost x2.0  cost x3.0
acc x0.4    18617.0      214.0        0.0        0.0        0.0
acc x0.7    86131.0    21943.0     2489.0        0.0        0.0
acc x1.0   143871.0    63133.0    26433.0     6039.0        0.0
acc x1.3   201120.0    99643.0    57040.0    29955.0     2132.0
acc x1.6   269323.0   130865.0    82231.0    54798.0    14684.0


05:59:20 | INFO    | model | most influential assumption: offer_cost_multiplier


05:59:20 | INFO    | model | saved reports/sensitivity.json + tornado_sensitivity.png


break-even acceptance scale     : 0.35
=> protection-bundle acceptance  : 12.2 % (baseline assumes 35%)


,offer_cost_multiplier,net_retained_inr
0,0.50,143871.0
1,0.75,94486.0
2,1.00,63133.0
3,1.50,26433.0
4,2.00,6039.0
5,3.00,0.0


![tornado](../reports/figures/tornado_sensitivity.png)

**Offer cost is the most influential assumption.** The campaign survives down to ~12.2% acceptance — a ~3× cushion under our 35% assumption.

## 5. What-if scenarios

In [7]:
import simulator as S
sc = S.run()
pd.DataFrame(sc['scenarios']).T[['n_contacted','spend_inr','net_retained_inr']]

05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
Empty DataFrame
Columns: [customers, avg_churn, avg_ltv, avg_net, total_cost]
Index: []


05:59:20 | INFO    | model | eligible (+EV under best offer): 0 / 7043


05:59:20 | INFO    | model | pessimistic  net=₹        0 | contacted=    0 | util=  0.0% | mix={}


05:59:20 | INFO    | model | Offer assignment (7,043 customers):


05:59:20 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:20 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:20 | INFO    | model | base         net=₹    63133 | contacted= 1349 | util=100.0% | mix={'Bill discount': 50, 'Courtesy call': 836, 'Protection bundle': 463}


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount           1561       0.56  1036.67   164.64    257565.0
Courtesy call           1262       0.29   947.24    14.32     37860.0
Protection bundle        571       0.56   781.74   105.59     51390.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 3394 / 7043


05:59:21 | INFO    | model | optimistic   net=₹   158925 | contacted=  741 | util=100.0% | mix={'Bill discount': 444, 'Protection bundle': 297}


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | Offer assignment (7,043 customers):


05:59:21 | INFO    | model | 
                   customers  avg_churn  avg_ltv  avg_net  total_cost
offer                                                                
Bill discount            147       0.70   935.47    73.42     32340.0
Courtesy call           1355       0.47   887.17    20.23     54200.0
Protection bundle        636       0.69   758.13    63.34     76320.0


05:59:21 | INFO    | model | eligible (+EV under best offer): 2138 / 7043


05:59:21 | INFO    | model | budget response:
 budget_inr  net_retained_inr  n_contacted  budget_utilisation_pct
      10000          10917.36          184                    99.2
      25000          24074.96          405                    99.7
      50000          40601.64          767                    99.8
      75000          53091.97         1062                    99.9
     100000          63132.68         1349                   100.0
     150000          77104.96         1845                   100.0
     200000          78490.00         2138                    81.4
     300000          78490.00         2138                    54.3


05:59:21 | INFO    | model | budget saturates near ₹200,000 — beyond this, extra budget buys nothing (no +EV customers left)


05:59:21 | INFO    | model | saved reports/scenarios.json


,n_contacted,spend_inr,net_retained_inr
pessimistic,0,0.0,0.0
base,1349,100000.0,63132.68
optimistic,741,99990.0,158924.89


**The pessimistic scenario contacts nobody.** Under weak acceptance, expensive offers and thin margin, *no* customer is positive-EV — so the system correctly recommends **not running the campaign**. A decision tool that knows when not to spend is more trustworthy than one that always does.

**Budget saturates near ₹2,00,000** — beyond that, extra budget buys nothing because no +EV customers remain.

*Next: `docs/AB_experiment_design.md` — how we'd replace the acceptance **assumption** with a **measurement**.*